File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:

        if len(strs) < 1 or "" in strs:
            return ""

        first = strs[0]
        for i in range(len(first)):
            for str in strs[1:]:
                if i == len(str) or first[i] != str[i]:
                    return first[0:i]
                    
        return first

'''
    "flower",
    "flow",
    "flight"

    brute force

    1. find LCP between first two
    2. find LCP between #1 and 3rd

    Trie

    f
    l 
    o i g t
    w- word
    e
    r - word
'''        

File: 0014-longest-common-prefix/0014-longest-common-prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.empty()) {
            return "";
        }
        for(const auto s : strs) {
            if(s.empty()) {
                return "";
            }
        }
        string first = strs[0];
        for(int i=0; i<first.size(); i++) {
            for(int j=1; j<strs.size(); j++) {
                if(i==strs[j].size() || first[i] != strs[j][i]) {
                    return first.substr(0, i);
                }
            }
        }
        return first;
    }
};